PIDToInheritApplicationStateFrom should be set by information already present in the GPU process

aestes opened the pull request:
#### 8b5a389b1c1c52b00da0cf8ea2f42f13e66f5274
<pre>
PIDToInheritApplicationStateFrom should be set by information already present in the GPU process
<a href="https://bugs.webkit.org/show_bug.cgi?id=297789">https://bugs.webkit.org/show_bug.cgi?id=297789</a>
<a href="https://rdar.apple.com/156277603">rdar://156277603</a>

Reviewed by Youenn Fablet and Jer Noble.

PIDToInheritApplicationStateFrom can be set by information already present in the GPU process
-- notably the UI process&apos; and per-page presenting application&apos;s PID. Removed code from the
WebContent process that would send a presenting application PID to the GPU process when an audio or
camera session will start and instead provided the presenting application PID to AVSystemController
at the time an AVAudioSession or AVCaptureSession starts in the GPU process.

For activating AVAudioSession on tvOS, the GPU process&apos; parent PID is used instead of the
presenting application PID to account for that platform&apos;s unique process model.

* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::start):
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::mediaSessionPresentingApplicationPID const): Deleted.
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaSessionPresentingApplicationPID const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setPresentingApplicationAuditToken):
* Source/WebCore/platform/audio/MediaSessionManagerInterface.h:
(WebCore::MediaSessionManagerInterface::prepareToSendUserMediaPermissionRequestForPage): Deleted.
(WebCore::MediaSessionManagerInterface::updatePresentingApplicationPIDIfNecessary): Deleted.
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.cpp:
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.h:
(WebCore::PlatformMediaSessionInterface::presentingApplicationPID const): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioSessionCocoa.mm:
(WebCore::AudioSessionCocoa::tryToSetActiveInternal):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.h:
(WebCore::MediaSessionManagerCocoa::providePresentingApplicationPIDIfNecessary): Deleted.
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
(WebCore::MediaSessionManagerCocoa::prepareToSendUserMediaPermissionRequestForPage): Deleted.
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.mm:
(MediaSessionHelperIOS::presentedApplicationPID const):
(MediaSessionHelperIOS::providePresentingApplicationPID):
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.mm:
(WebCore::MediaSessionManageriOS::sessionWillBeginPlayback):
(WebCore::MediaSessionManageriOS::providePresentingApplicationPIDIfNecessary): Deleted.
(WebCore::MediaSessionManageriOS::updatePresentingApplicationPIDIfNecessary): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::ensureMediaSessionHelper):
(WebKit::GPUConnectionToWebProcess::overridePresentingApplicationPIDIfNeeded): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
(WebKit::RemoteAudioSessionProxyManager::tryToSetActiveForProcess):
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.h:
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.messages.in:
* Source/WebKit/GPUProcess/webrtc/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::startProducingData):
* Source/WebKit/GPUProcess/webrtc/UserMediaCaptureManagerProxy.h:
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::startProducingData):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::providePresentingApplicationPID): Deleted.
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:

Canonical link: <a href="https://commits.webkit.org/299129@main">https://commits.webkit.org/299129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5c5889bf24bf971b1e82ebda9bd77467a501247

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117941 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28254 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124088 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69976 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d3092c39-f117-487e-96e4-83d286828542) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119819 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38312 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46201 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89502 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59113 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3646d6ea-3c2c-4068-bf9b-5ed077e835bc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120893 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30509 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105756 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69996 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bcfbd259-5697-47ab-8ae0-9837a3ff6481) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/117301 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29576 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23871 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67751 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99931 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24050 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127169 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44844 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33781 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98171 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45205 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101980 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97959 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24920 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43353 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21335 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41278 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44716 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50390 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44176 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47521 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45865 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2fd7eca7eeb466a386af1ffd846de547c3cc3c63. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49775

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/01c8abc3386f3dfcefe0cd814af851b0f56810c4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49775

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fb01bce5672f4e6443429e2dd1c922854bd3980d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49775

youennf commented on Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.h:
```diff
@@ -91,9 +92,8 @@ class WEBCORE_EXPORT MediaSessionHelper : public ThreadSafeRefCountedAndCanMakeT
-    void providePresentingApplicationPID(int pid) { providePresentingApplicationPID(pid, ShouldOverride::No); }
-    virtual void providePresentingApplicationPID(int, ShouldOverride) = 0;
+    virtual std::optional<ProcessID> presentedApplicationPID() const;
+    virtual void providePresentingApplicationPID(ProcessID);
```
Do we even need it to be virtual anymore since it is no longer implemented by the remote helper? Maybe as a follow-up?

youennf commented on Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.mm:
```diff
@@ -152,8 +130,6 @@
     session.setShouldPlayToPlaybackTarget(playbackTargetSupportsAirPlayVideo);
 #endif
 
-    providePresentingApplicationPIDIfNecessary(session.presentingApplicationPID());
```
In the case of a muted video, we would no longer set the presenting application PID since the remote audio session would not be activated. I don't think there is any user impact though.

youennf commented on Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
```diff
@@ -188,6 +224,9 @@ bool RemoteAudioSessionProxyManager::tryToSetActiveForProcess(RemoteAudioSession
         // This proxy and only this proxy wants to become active. Activate
         // the session, and return whether that activation was successful.
+#if PLATFORM(IOS_FAMILY)
+        providePresentingApplicationPID(proxy);
```
I wonder what happens in the following use case:
- A first presenting app/web page using GPUProcess G starts camera/microphone capture (or maybe audio context in media playback session type).
- A second presenting app/web page using G starts an AudioContext. It wants to activate the audio session in ambient.
- The first presenting app/web page may or may not be interrupted.

The presenting application PID will likely stay to the one of the first app, while I think it would move to the second app before the PR. I am not sure whether we want to handle this case.

youennf reviewed (commented):
Looks like a good simplification.
A few thoughts below (PR is still in draft btw).

aestes commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5872,14 +5872,6 @@ const std::optional<audit_token_t>& Page::presentingApplicationAuditToken() cons
 void Page::setPresentingApplicationAuditToken(std::optional<audit_token_t> presentingApplicationAuditToken)
 {
     m_presentingApplicationAuditToken = WTFMove(presentingApplicationAuditToken);
```
Probably not. I will remove this in a follow-on.

jernoble reviewed (approved):
So much red! r=me.

youennf reviewed (approved):
LGTM

aestes committed:
PIDToInheritApplicationStateFrom should be set by information already present in the GPU process
https://bugs.webkit.org/show_bug.cgi?id=297789
rdar://156277603

Reviewed by Youenn Fablet and Jer Noble.

PIDToInheritApplicationStateFrom can be set by information already present in the GPU process
-- notably the UI process' and per-page presenting application's PID. Removed code from the
WebContent process that would send a presenting application PID to the GPU process when an audio or
camera session will start and instead provided the presenting application PID to AVSystemController
at the time an AVAudioSession or AVCaptureSession starts in the GPU process.

For activating AVAudioSession on tvOS, the GPU process' parent PID is used instead of the
presenting application PID to account for that platform's unique process model.

* Source/WebCore/Modules/mediastream/UserMediaRequest.cpp:
(WebCore::UserMediaRequest::start):
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::mediaSessionPresentingApplicationPID const): Deleted.
* Source/WebCore/Modules/webaudio/AudioContext.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::mediaSessionPresentingApplicationPID const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setPresentingApplicationAuditToken):
* Source/WebCore/platform/audio/MediaSessionManagerInterface.h:
(WebCore::MediaSessionManagerInterface::prepareToSendUserMediaPermissionRequestForPage): Deleted.
(WebCore::MediaSessionManagerInterface::updatePresentingApplicationPIDIfNecessary): Deleted.
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.cpp:
* Source/WebCore/platform/audio/PlatformMediaSessionInterface.h:
(WebCore::PlatformMediaSessionInterface::presentingApplicationPID const): Deleted.
* Source/WebCore/platform/audio/cocoa/AudioSessionCocoa.mm:
(WebCore::AudioSessionCocoa::tryToSetActiveInternal):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.h:
(WebCore::MediaSessionManagerCocoa::providePresentingApplicationPIDIfNecessary): Deleted.
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
(WebCore::MediaSessionManagerCocoa::prepareToSendUserMediaPermissionRequestForPage): Deleted.
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.mm:
(MediaSessionHelperIOS::presentedApplicationPID const):
(MediaSessionHelperIOS::providePresentingApplicationPID):
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionManagerIOS.mm:
(WebCore::MediaSessionManageriOS::sessionWillBeginPlayback):
(WebCore::MediaSessionManageriOS::providePresentingApplicationPIDIfNecessary): Deleted.
(WebCore::MediaSessionManageriOS::updatePresentingApplicationPIDIfNecessary): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::ensureMediaSessionHelper):
(WebKit::GPUConnectionToWebProcess::overridePresentingApplicationPIDIfNeeded): Deleted.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/RemoteAudioSessionProxyManager.cpp:
(WebKit::RemoteAudioSessionProxyManager::tryToSetActiveForProcess):
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.h:
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.messages.in:
* Source/WebKit/GPUProcess/webrtc/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::startProducingData):
* Source/WebKit/GPUProcess/webrtc/UserMediaCaptureManagerProxy.h:
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::startProducingData):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::providePresentingApplicationPID): Deleted.
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:

Canonical link: https://commits.webkit.org/299129@main
